src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs

```diff
@@ -96,12 +97,107 @@ public static void Fill(Span<byte> data)
             RandomNumberGeneratorImplementation.FillSpan(data);
         }
 
+        public static int GetInt32(int fromInclusive, int toExclusive)
+        {
+            if (fromInclusive >= toExclusive) throw new ArgumentException(SR.Argument_InvalidRandomRange);
+
+            // The total possible range is [0, 4,294,967,295).
+            long adjustedRange = (long)toExclusive - fromInclusive;
+
+            // If there is only one possible choice, nothing random will actually happen, so return
+            // the only possibility.
+            if (adjustedRange == 1)
+            {
+                return fromInclusive;
+            }
+
+            // Subtract one to account for zero being an actual possibility. We know for certain that
+            // the range fits in a UInt32 without overflowing.
+            uint maskedRange = (uint)adjustedRange - 1;
+
+            int leadingZeroCount = LeadingZeroCount(maskedRange);
+
+            // Create a mask for the bits that we care about for the range. The other bits will be
+            // masked away.
+            uint mask;
+
+            // If the LZCNT is zero create a full mask to avoid overflowing when shifting.
+            if (leadingZeroCount == 0)
+            {
+                mask = unchecked((uint)-1);
+            }
+            else
+            {
+                mask = (1U << 32 - leadingZeroCount) - 1;
+            }
+
+            // We only want to generate as many bytes as required to satisfy the mask to not apply
```

Comment:
You can also get away with not relying on the lzcnt instruction, since if you're draining 4 bytes at a time you don't really care about the leading zero count. All you really care about is smearing the most significant set bit rightward to generate a mask:

```cs
// assume 'value' is uint
uint mask = value;
mask |= mask >> 1;
mask |= mask >> 2;
mask |= mask >> 4;
mask |= mask >> 8;
mask |= mask >> 16;
```

Reply:
>Is it really that bad to drain 4 bytes at a time from the CSPRNG? Seems like it would also simplify the logic below considerably.

I don't really have a strong opinion - agree the code is simpler, but my overall feeling is that while it adds a lot of lines of code, it doesn't add that much additional complexity.

@bartonjs ?

Reply:
For context, I think the majority of random numbers are for indexing into collections to pick a random "thing", and most collections are fairly small, less than 256, so we'd be asking the CSPRNG for something that we throw away 75% of the time. I know the Windows RNG is pretty capable of dealing with that, but I'm not sure about other platforms like IoT where it may have a larger impact.

Reply:
@GrabYourPitchforks  You've worked with much more constrained environments than I have.  If you don't think that burning that much data from the CSPRNG is bad then I'll endorse the "simple until we need to be more complicated" plan.

All of our "with a keyboard" systems are probably measured in MB/s, but I don't have a sense of the limit of MD5 calls per second on a Raspberry Pi.  (Or whatever PRF is used these days)

Reply:
I'm not aware of any device where it's significantly more expensive to drain 4 bytes from the RNG than it is to drain fewer than 4 bytes. Maybe we'll have such a platform in the future. For now IMO our best bet is to keep the code as simple as possible, and we can optimize in a future release if a new requirement comes online.

Reply:
Okie doke, simple it is.